Some headers were missing when serving from cache

Chraneco opened the pull request:
The setting of headers like 'Content-Type' is currently happening in method `respond` of class `JApplicationWeb`.

Unfortunately, the system cache plugin stores contents at event `onAfterRender` which is triggered before `respond` is executed. Therefore, certain headers are missing when serving from the cache.
This creates problems especially if we are serving content that is not of Content-Type `text/html`.

This pull request changes the plugin so that the cache is created only after these headers are set by using event `onAfterRespond` instead of `onAfterRender`.

Since gzip compression already happened at this point, the compressed contents are stored in the cache and therefore no further compression is necessary when serving cached contents. Thus, the second change in the plugin file.

For testing please make sure that everything works as before and check that the `Content-Type` header is also cached now.
### Testing Instructions
1. Install attached cache test component and access it in frontend (there's no backend part).
2. Click on the presented link which takes you to an image which should be displayed fine (even after refreshing the page).
3. Enable the cache plugin of Joomla! and ensure that debug mode is disabled and you aren't logged in in frontend.
4. Refresh the page of the image a few times. => It won't be displayed correctly anymore.
5. Apply the patch, clear the cache and do the same again => the image will be displayed even after refreshing.

Chraneco committed:
Some headers were missing when serving from cache

RickR2H commented:
hello @Chraneco

Thank you for your contribution.

Please provide clear test instructions to be able to test / reproduce this issue.
If no reply is received within 4 weeks we will close this issue.

Thanks for understanding!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/6955">issues.joomla.org/joomla-cms/6955</a>.</sub>

Chraneco commented:
Yes, I added instructions above.

Unfortunately, it doesn't seem to be possible to add extensions archives as attachments here, so I added an issue on Joomlacode and attached the test component there:
http://joomlacode.org/gf/project/joomla/tracker/?action=TrackerItemEdit&tracker_item_id=34149&start=0<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/6955">issues.joomla.org/joomla-cms/6955</a>.</sub>

RickR2H commented:
@Chraneco I tried the component but initially the image wouldn't show. I do get the link but when I click it I get an empty page... Hope you can fix the component.
[afbeelding 4](https://cloud.githubusercontent.com/assets/5610413/9314888/0840a6c2-452c-11e5-9158-748d542634d6.jpg)

Chraneco commented:
@RickR2H Sorry, the sample image in Joomla I was using was apparently deleted in one of the latest versions. I updated the component with a path to another image and sent it to your email address.

If it still doesn't work please check whether the image `images/sampledata/fruitshop/apple.jpg` exists in your Joomla installation (it's there in a fresh new instance).

Thanks for testing!

RickR2H commented:
Patch works! Thanks @Chraneco for putting all the effort in making a component to test this PR. <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/6955">issues.joomla.org/joomla-cms/6955</a>.</sub>

RichardR2H commented:
Works here! Thanks.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/6955">issues.joomla.org/joomla-cms/6955</a>.</sub>

zero-24 commented:
RTC based on tests by @RickR2H and @RichardR2H.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/6955">issues.joomla.org/joomla-cms/6955</a>.</sub>
